### PR TITLE
Adopt to the breaking changes in WordPressOrgRestApi

### PR DESCRIPTION
**Pull request by crazytonyli**
## Description

`WordPressOrgRestApi` used to only support self-hosted sites. It's been rewritten to support self hosted sites and WP.com sites. This PR adopts this breaking changes.

The existing `Blog.wordPressOrgRestApi` returns a non-nil reference only when the blog is a self-hosted site, which is not obvious from the property name "wordPressOrgRestApi", because every WordPress site is bundled with [.org rest API](https://developer.wordpress.org/rest-api/). I have renamed it to `selfHostedSiteRestApi`, especially considering `WordPressOrgRestApi` now supports all sites.

## Test Instructions

There are only a number of places in the app that calls .org REST API. Editor settings API is one of them, which will be fired upon entering Pages List and opening a post. Plugin is another place that uses .org REST API, which is only accessible using self hosted site in the app.

You can put a breakpoint in `BlockEditorSettingsService.fetchTheme` function to see the API result. Or, you can use a proxy to see all 'wp/v2' HTTP requests.

## Regression Notes
1. Potential unintended areas of impact

All features that call .org REST API. i.e. plugins and gutenberg settings.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Verified plugins and gutenberg settings API calls work.

3. What automated tests I added (or what prevented me from doing so)

None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist: N/A

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Podfile reference(s) to a branch:<br />
<code>WordPressKit (from `https://github.com/wordpress-mobile/WordPressKit-iOS.git`, branch `wordpress-org-rest-api-v2`)</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by mokagio** [comment]
One of the UI tests [failures](https://buildkite.com/automattic/wordpress-ios/builds/20452#018dbe88-2fc5-4998-ad35-1b455f547b59) occurred in the Gallery block.

For some reason, the "Content Structure" that the test runner read returned 2 blocks instead of the expected 5. However, if you look at the test artifacts, you can see the 5 blocks on screen:

image

**Comment by mokagio** [comment]
@crazytonyli I'm running out of time in my day. I was hoping to be able to fix the issue and review these changes, but I'm afraid I'll then have to rush through the rest of the [24.3](https://github.com/wordpress-mobile/WordPress-iOS/milestone/271) code freeze.

As such, I'll push this to the next milestone and do as you suggested and release a stable WordPressKit version that doesn't include the `trunk` changes this PR points to.

This will delay testing the new code IRL, but it will give us time to properly investigate all issues and review the PR. 

Thanks 🙌

**Comment by crazytonyli** [comment]
@mokagio That sounds good to me!

**Comment by crazytonyli** [comment]
I have spent quite a while trying to compare the trunk version and this PR version of the app to understand what's the correct behaviour. But I probably don't know Gutenberg Editor enough to understand how to count blocks. 

https://github.com/wordpress-mobile/WordPress-iOS/blob/f0755290a7a16c24294290675ddd619142717e7f/WordPress/UITests/Tests/EditorGutenbergTests.swift#L94-L100

However, I have perform the steps in this test case on the trunk branch on a freshly installed app, the number of blocks I get is 5, not 2 as asserted by the UI test.

https://github.com/wordpress-mobile/WordPress-iOS/assets/1101828/ee8c2618-f739-45ae-ab9f-d3ad2443685e

But the UI test on the trunk branch passes, which means the UI test case is correct. Does that mean, we get different results on the actual app and the UI test app?

@wordpress-mobile/mobile-ui-testing-squad can you please help me to understand how to verify the correctness of `testAddGalleryBlock`?

**Comment by tiagomar** [comment]
Hey @crazytonyli 👋 

I couldn't investigate the block count failure in this branch because I haven't been able to build. Not sure if I'm missing something, but I get PluginStore build erros:

image

>  Does that mean, we get different results on the actual app and the UI test app?

Yeah, that's true for `trunk` right now and I don't know why it's different. 🤔 
Looking into the Post's HTML version of similar posts from the JP and JP UI Tests apps we can spot the difference. While the images inside the Gallery block are represented as individual blocks on the JP app, there's only the Gallery block on JP UI Tests app. Please check the image below for reference.

image

A visual hint that it's really different is the focused block after the images insertions: a single image for the JP app and the Gallery block for JP UI Tests app.

image

I've tested it on a real device on 24.2 and 23.7 and in both cases the count was the same as in `trunk`, 5 blocks.

From the image in @mokagio's comment, the first image is in focus and the blocks count was 5 (which is in line with what we observe in real devices and for the JP app) while the UI Tests expected 2 blocks (1 Paragraph and 1 Gallery). So, I still don't understand why, but it looks like something in this branch migh have fixed the JP and JP Unit Tests discrepancy? 🤔  In this case we should just update the test to `.verifyContentStructure(blocks: 5, words: postContent.components(separatedBy: " ").count, characters: postContent.count)`.

File: WordPress/Classes/Services/BlockEditorSettingsService.swift [hunk 1]
```diff
@@ -22,22 +22,15 @@ class BlockEditorSettingsService {
     }
 
     convenience init?(blog: Blog, coreDataStack: CoreDataStackSwift) {
-        let remoteAPI: WordPressRestApi
-        if blog.isAccessibleThroughWPCom(),
-           blog.dotComID?.intValue != nil,
-           let restAPI = blog.wordPressComRestApi() {
```
Comment (crazytonyli) [inline WordPress/Classes/Services/BlockEditorSettingsService.swift:28]: @mokagio @tiagomar I think I have found the root cause of the `testAddGalleryBlock` failure: .org REST API calls stopped using API-Mocks in this PR. We used to use a `WordPressComRestApi` instance, which can be mocked by UI tests API-Mocks. But the new `WordPressOrgRestApi` doesn't support that. I'll just need to add that support back by allowing passing a custom "api base URL", which can be set to mock server url during UI tests.

Here are some details about the issue.

As @tiagomar has pointed out, the "html content" is different: one use images block and the other use a gallery block. [This variable here](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Models/BlockEditorSettings%2BGutenbergEditorSettings.swift#L14) dictates which block to use. It's part of [a protocol defined in Gutenberg Mobile](https://github.com/WordPress/gutenberg/blob/3888f1e58528be5a641621ab171eba9f1e816928/packages/react-native-bridge/ios/GutenbergBridgeDataSource.swift#L85). One place, which appears to be the only place in UI tests, that makes Gutenberg Mobile calls this variable is [_successfully_ fetching and updating editor settings](https://github.com/wordpress-mobile/WordPress-iOS/blob/ce3fea5762dba5b6ceac1e09904f39ea1981f7f3/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L1431), which sends a .org REST API call. As I mentioned before, this PR breaks API mocks for .org REST API calls, which results in a fetching editor settings failure, which ultimately leads the editor produces a different html content.

Comment (crazytonyli) [reply WordPress/Classes/Services/BlockEditorSettingsService.swift:28]: Should be fixed in https://github.com/wordpress-mobile/WordPressKit-iOS/pull/731

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/Services/BlockEditorSettingsService.swift (hunk 1 above)
Comment (mokagio) [reply WordPress/Classes/Services/BlockEditorSettingsService.swift:28]: > But the new `WordPressOrgRestApi` doesn't support that. I'll just need to add that support back by allowing passing a custom "api base URL", which can be set to mock server url during UI tests.

I think being able to customize the base URL is good design regardless, even if 99% of the usages will be with the default one configured in the library.

**Comment by crazytonyli** [comment]
@tiagomar Going back to the discrepancy in Gutenberg Editor between using an actual site and the test site used in UI tests, I think it might be good to consider "updating" the test site to an more up-to-date version. [The test site is on "5.2.2"](https://github.com/wordpress-mobile/WordPress-iOS/blob/730306f2583f90d25acd9893e53d4da70779f59f/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/sites/106707880%20-%20Tri-County%20Real%20Estate%20-%20Personal%20Plan/rest_sites_106707880.json#L103), which I believe is why it shows "2 blocks" in this particular test case. Changing it to "6.4" or something (I'm not sure if [the rest of the site data](https://github.com/wordpress-mobile/WordPress-iOS/blob/730306f2583f90d25acd9893e53d4da70779f59f/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/sites/106707880%20-%20Tri-County%20Real%20Estate%20-%20Personal%20Plan/rest_sites_106707880.json#L13) needs to be changed too) probably would make Gutenberg Editor use 5 blocks like on a non-UI-test app.

**Review by mokagio — APPROVED** [review]

**Comment by mokagio** [comment]
I'm trying to see how many of the types in WordPressKit closer to the app domain I can move out of the library and into this repo, given WooCommerce doesn't directly depend on them, see https://github.com/woocommerce/woocommerce-ios/pull/12063.

Having these changes merged will help keep my Git history tidy (tidier)

**Pull request merged**